### PR TITLE
SamLocusAndReferenceIterator fails with leading indels at the start of the reference

### PR DESCRIPTION
**Pull request by nh13**
This comes from https://github.com/broadinstitute/picard/issues/1511

### CODE REVIEW

**Comment by nh13** [comment]
@yfarjoun I haven't contributed in a while, but could you take a look, or assign someone who can?

File: src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java [hunk 1]
```diff
@@ -53,4 +53,52 @@ public void testSamLocusAndReferenceIteratorMismatch() {
...
+        int totalDeletedRecords = 0;
+        for (final SamLocusAndReferenceIterator.SAMLocusAndReference samLocusAndReference : samLocusAndReferences) {
+
+            totalMappedRecords += samLocusAndReference.getRecordAndOffsets().size();
+            totalInsertedRecords += samLocusAndReference.getLocus().getInsertedInRecord().size();
+            totalDeletedRecords += samLocusAndReference.getLocus().getDeletedInRecord().size();
+
+            if (samLocusAndReference.getLocus().getInsertedInRecord().size() > 0) {
+                System.err.println(samLocusAndReference.getLocus());
+            }
```
Comment (nh13) [inline src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java:90]: oopsie, this needs to be removed

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 2]
```diff
@@ -78,8 +78,12 @@ public SAMLocusAndReference next() {
         final ReferenceSequence referenceSequence = referenceSequenceFileWalker.get(locus.getSequenceIndex(), locus.getSequenceName(),
                 locus.getSequenceLength());
 
-        //position is 1-based...arrays are 0-based!
-        return new SAMLocusAndReference(locus, referenceSequence.getBases()[locus.getPosition() - 1]);
+        // Developer notes:
+        // 1. position is 1-based...arrays are 0-based!
+        // 2. We must guard against insertions before the reference
+        final int position = locus.getPosition();
+        final byte referenceBase = 0 == position ? (byte)'N' : referenceSequence.getBases()[locus.getPosition() - 1];
```
Comment (tfenne) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java:85]: Would you mind a) adding a `final byte N = (byte) 'N'` and referencing it here.  Could you also please wrap `0 == position` in parentheses?  I think that'll make the ternary easier to read.

File: src/test/resources/htsjdk/samtools/reference/leading_insertion.sam [hunk 3]
```diff
@@ -0,0 +1,8 @@
+@HD	VN:1.0	SO:coordinate
+@SQ	SN:chrM	LN:16571	UR:file:/Users/mhanna/src/Sting/Homo_sapiens_assembly18.trimmed.fasta	M5:d2ed829b8a1628d16cbeee88e88e39eb
+@SQ	SN:chr20	LN:1000000	UR:file:/Users/mhanna/src/Sting/Homo_sapiens_assembly18.trimmed.fasta	M5:b4eac854d70893986ac578c53c2324f1
+@RG	ID:62A40.2	PL:illumina	PU:62A40AAXX101028.2	LB:Solexa-45345	DT:2010-10-28T00:00:00-0400	SM:RRBS885	CN:BI
+1	99	chrM	1	60	10I26M	chrM	1	0	AAAAAAAAAANNNNNNNNNNNNNNNNNNNNNNNNNN	AAAAAAAAAAAAAAABBBBBBBBBBBBBBBBBBBBB	RG:Z:62A40.2
```
Comment (tfenne) [inline src/test/resources/htsjdk/samtools/reference/leading_insertion.sam:5]: Would it be possible to use real bases here? E.g. the last 10bp of chrM concatenated with the first 26 (I assume that's what's causing these insertions?).

File: src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java [hunk 4]
```diff
@@ -53,4 +53,52 @@ public void testSamLocusAndReferenceIteratorMismatch() {
...
+        overlapDetector.addAll(intervalList.getIntervals(), intervalList.getIntervals());
+
+        // Developer note: these are the # of loci that we have (1) a read base mapped to a reference base, (2) a read
+        // base inserted relative the reference, and (3) a read base deleted relative to the reference.  For read bases
+        // inserted prior to or after the reference (ex.. at position 1 of a contig), they are counted only once!
+        int totalMappedRecords = 0;
+        int totalInsertedRecords = 0;
+        int totalDeletedRecords = 0;
+        for (final SamLocusAndReferenceIterator.SAMLocusAndReference samLocusAndReference : samLocusAndReferences) {
+
```
Comment (tfenne) [inline src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java:83]: Could you also assert that you see reference position `0` and that the reference base is `N` at that position please?

**Comment by nh13** [comment]
@lbergelson can you make a release with this so I can update Picard?

**Comment by lbergelson** [comment]
@nh13 Yes.  I want to do one very soon for other reasons as well.  Can you wait a few days until I get a few other things merged?

**Comment by nh13** [comment]
Excellent, a couple days would be great!!!

**Comment by nh13** [comment]
@lbergelson welcome back from holiday, do you think you'll be able to get to it this week?  I see the last release [2.21.3](https://github.com/samtools/htsjdk/releases/tag/2.21.3) was Mar 9, so there may well be a good set of other changes to include too.

**Comment by lbergelson** [comment]
Thanks.  My hope is to get it out in the next day or two.

**Comment by lbergelson** [comment]
There are some changes in the pipe that people also want out but probably won't get merged by then so we'll probably do another one pretty quickly.

**Comment by lbergelson** [comment]
@nh13 2.22.0 is released and should be available in central soon

**Pull request merged**